### PR TITLE
Defer loading SolidityContract

### PR DESCRIPTION
**Pull request by dkent600**
Addresses: 

- https://github.com/daostack/arc.js/issues/130
- https://github.com/daostack/arc.js/issues/133

Defer loading SolidityContract
Renamed getScheme to getContractWrapper
Bumped version number
Pull founders from json file at deployment

**Breaking changes**: 
- Renamed getScheme to getContractWrapper
- Removed UController from `Contracts.getDeployedContracts`  (I really need to rename `Contracts.getDeployedContracts`)

### CODE REVIEW

File: lib/utils.ts [hunk 1]
```diff
@@ -42,12 +42,21 @@ export class Utils {
...
    */
-  public static getWeb3(): any {
-    if (Utils.web3) {
+  public static getWeb3(forceReload: boolean = false): any {
+
+    if (forceReload) {
+      delete Utils.web3;
+      Utils.alreadyTriedAndFailed = false;
+      if (typeof window !== "undefined") {
+        delete window.web3;
```
Comment (tibetsprague) [inline lib/utils.ts]: Hmm. Won’t this eliminate access to meta mask version of web3? I think we can reload web3 a different way. Maybe by calling get accounts?

Comment (dkent600) [reply lib/utils.ts]: We set window.web3 below.

Comment (dkent600) [reply lib/utils.ts]: Having said  that, I do question how it would affect the following ine:

`if (typeof web3 !== "undefined")`

Comment (dkent600) [reply lib/utils.ts]: I'm not clear how "calling get accounts" would work.  Can you describe?

File: migrations/2_deploy_organization.js [hunk 2]
```diff
@@ -30,50 +31,48 @@ module.exports = async (deployer) => {
...
-  const initRepInWei = [web3.toWei(10)];
-  const initTokenInWei = [web3.toWei(1000)];
   const orgNativeTokenFee = 0;
   const defaultVotingMachineParams = {
     preBoostedVoteRequiredPercentage: 50,
-    preBoostedVotePeriodLimit: 60,
-    boostedVotePeriodLimit: 60,
-    thresholdConstA: 1,
-    thresholdConstB: 1,
+    preBoostedVotePeriodLimit: 5184000, // 2 months
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: from where these params come from ?

File: migrations/2_deploy_organization.js [hunk 3]
```diff
@@ -30,50 +31,48 @@ module.exports = async (deployer) => {
...
+    boostedVotePeriodLimit: 604800, // 1 week
+    thresholdConstA: 1, // TODO: how many proposals can be boosted at a time. want about 5, have to read arc to figure out
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: How this param is related to "how many proposals can be boosted at a time"?

File: migrations/2_deploy_organization.js [hunk 4]
```diff
@@ -30,50 +31,48 @@ module.exports = async (deployer) => {
...
+    thresholdConstB: 1, // TODO: exponent of how hard to get the next one
     minimumStakingFee: 0,
-    quietEndingPeriod: 0,
-    proposingRepRewardConstA: 1,
-    proposingRepRewardConstB: 1,
-    stakerFeeRatioForVoters: 1,
-    votersReputationLossRatio: 10,
-    votersGainRepRatioFromLostRep: 80,
+    quietEndingPeriod: 7200, // Two hours
+    proposingRepRewardConstA: 5, // baseline rep rewarded TODO: good for now but Adam going to look up what this should be
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: I would remove personal names .
It should be 

> baseline rep rewarded TODO: figure out the real default value.

File: migrations/2_deploy_organization.js [hunk 5]
```diff
@@ -30,50 +31,48 @@ module.exports = async (deployer) => {
...
       tokenSymbol,
-      founders,
-      initTokenInWei,
-      initRepInWei,
-      // use the universal controller
-      0);
+      founders.map((f) => f.address),
+      founders.map((f) => web3.toWei(f.tokens)),
+      founders.map((f) => web3.toWei(f.reputation)),
+      // don't use the universal controller
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: will change that to : 
> use single controller .

File: migrations/founders.kovan.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "founders": [
```
Comment (orenyodfat) [inline migrations/founders.kovan.json]: please add a comment that this is for testing purpose only .

Comment (orenyodfat) [reply migrations/founders.kovan.json]: why do we need to forge with founders for kovan ?Is that only for test ? 
Is that for alchemy on Kovan ? If yes, does some one has the prv keys for these  addresses ?

File: migrations/founders.kovan.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (orenyodfat) [inline migrations/founders.kovan.json]: do we need kovan as part of the file name ?

File: migrations/2_deploy_organization.js [hunk 8]
```diff
@@ -30,50 +31,48 @@ module.exports = async (deployer) => {
   const orgName = "Genesis";
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: Will change the name to "Genesis-Test" or something which defer that from "Genesis"

File: lib/utils.ts [hunk 9]
```diff
@@ -42,9 +42,10 @@ export class Utils {
...
 
   /**
-   * Returns the web2 object.
+   * Returns the web3 object.
    * When called for the first time, web3 is initialized from the Arc.js configuration.
-   * Throws an exception when web3 cannot be initialized or there is no default client
+   * Throws an exception when web3 cannot be initialized.
+   * @param {boolean} forceReload true to reload/retry web3
    */
   public static getWeb3(): any {
```
Comment (orenyodfat) [inline lib/utils.ts]: why this function is needed ?

File: migrations/2_deploy_organization.js (hunk 2 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: They come from Adam and Tiber, as far as I know.  They are not complete.

File: migrations/founders.ganache.json [hunk 10]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (orenyodfat) [inline migrations/founders.ganache.json]: why need this file ?
why need ganache in the file name ?
why not just have founders.json for both ganache and kovan ?

File: migrations/2_deploy_organization.js (hunk 3 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: @leviadam @tibetsprague

File: migrations/2_deploy_organization.js (hunk 4 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: these are not complete

File: migrations/2_deploy_organization.js (hunk 2 above)
Comment (orenyodfat) [reply migrations/2_deploy_organization.js]: @leviadam ?

File: migrations/founders.kovan.json (hunk 7 above)
Comment (dkent600) [reply migrations/founders.kovan.json]: there is also one with ganache in the name

Comment (orenyodfat) [reply migrations/founders.kovan.json]: why not have just founders.json (for both ganache and kovan ).

Comment (orenyodfat) [reply migrations/founders.kovan.json]: for all networks

File: migrations/2_deploy_organization.js (hunk 2 above)
Comment (tibetsprague) [reply migrations/2_deploy_organization.js]: yup, Adam gave them to me, still gotta figure out the threshold params

File: migrations/founders.kovan.json (hunk 6 above)
Comment (tibetsprague) [reply migrations/founders.kovan.json]: this is why I dont think this file should be in the repo, we should be able to pass in a founders file if we want

Comment (tibetsprague) [reply migrations/founders.kovan.json]: yes this is to setup the genesis DAO how we want it setup on deployment

File: migrations/2_deploy_organization.js (hunk 8 above)
Comment (tibetsprague) [reply migrations/2_deploy_organization.js]: ? this is the real genesis DAO, especially when we deploy to main net

File: migrations/2_deploy_organization.js (hunk 3 above)
Comment (tibetsprague) [reply migrations/2_deploy_organization.js]: * @dev threshold return the organization's score threshold which required by
     * a proposal to shift to boosted state.
     * This threshold is dynamically set and it depend on the number of boosted proposal.

Comment (tibetsprague) [reply migrations/2_deploy_organization.js]: ok so @fmatan suggests thresholdConstA: 2, thresholdConstB: 10, to start

File: migrations/founders.kovan.json (hunk 6 above)
Comment (orenyodfat) [reply migrations/founders.kovan.json]: I agree with @tibetsprague  I feel that the genesisParams + the genesis_test founders + genesis dao should not be part of the arc.js .there should be passed as params via alchemy ...(manually or via file).

@tibetsprague I guess that this setup is not for the real final genesis DAO . it is just for demo. 
and it should be comment and maybe also should be reflected in the dao name.

Comment (tibetsprague) [reply migrations/founders.kovan.json]: hmm, yeah I guess we could deploy without any DAOs and setup the Genesis DAO manually. but its nice to have it setup for us too for dev and testing

Comment (orenyodfat) [reply migrations/founders.kovan.json]: "but its nice to have it setup for us too for dev and testing" 
Yes , but you can have it by default on alchemy (not arc.js)..no ?

Comment (tibetsprague) [reply migrations/founders.kovan.json]: how so? deployment happens in Arc.js. I mean its nice having a DAO setup on deployment, so we dont have to do it manually each time.

File: migrations/founders.ganache.json (hunk 10 above)
Comment (dkent600) [reply migrations/founders.ganache.json]: merged both into founders.json

File: migrations/2_deploy_organization.js (hunk 8 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: @orenyodfat can you please address Tibet's question, thanks!

File: migrations/founders.json [hunk 11]
```diff
@@ -1,13 +1,33 @@
 {
   "founders": {
     "ganache": [
+      {
```
Comment (orenyodfat) [inline migrations/founders.json]: why you need to mention the  'ganache' ?
founders are founders are founders .

File: migrations/2_deploy_organization.js [hunk 12]
```diff
@@ -6,9 +6,8 @@ module.exports = async (deployer) => {
 
   const gasAmount = Config.get("gasLimit_deployment");
   const network = Config.get("network");
-  console.log(`Deploying to ${network}, gasLimit: ${gasAmount}`);
-
   const founders = require("./founders.json").founders[network];
+  console.log(`Deploying to ${network}, gasLimit: ${gasAmount},  ${founders.length} founders`);
```
Comment (orenyodfat) [inline migrations/2_deploy_organization.js]: please remove console.log

Comment (orenyodfat) [reply migrations/2_deploy_organization.js]: how does it pass the eslint ?

File: migrations/founders.json (hunk 11 above)
Comment (orenyodfat) [reply migrations/founders.json]: @dkent600 ?

File: migrations/2_deploy_organization.js (hunk 12 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: this is a script that logs lots of information on the console as a normal and useful part of its function.  I would prefer to leave this in.  Does that work for you?

File: migrations/founders.json [hunk 13]
```diff
@@ -0,0 +1,83 @@
+{
+  "founders": {
+    "ganache": [
+      {
+        "address": "0x8d4408014d165ec69d8cc9f091d8f4578ac5564f376f21887e98a6d33a6e3549",
```
Comment (orenyodfat) [inline migrations/founders.json]: Why these adrresses are longer that kovan addresses ?
You can use the same address for kovan and ganache..no ?

File: migrations/founders.json (hunk 11 above)
Comment (dkent600) [reply migrations/founders.json]: This file contains both ganache and kovan founder data.

Comment (orenyodfat) [reply migrations/founders.json]: why ?

File: migrations/founders.json (hunk 13 above)
Comment (dkent600) [reply migrations/founders.json]: these are private keys that ganache seems to like.  they correspond to the accounts that we tell ganache to create when we fire it up.

Comment (dkent600) [reply migrations/founders.json]: The kovan addresses correspond to accounts on kovan that won't exist on ganache.

Comment (orenyodfat) [reply migrations/founders.json]: - please replace private key with address .
- if it is not possible  change : "address" tp "private key" or at least comment that.

File: migrations/2_deploy_organization.js (hunk 12 above)
Comment (orenyodfat) [reply migrations/2_deploy_organization.js]: no.how the linter did not catch that ?

File: migrations/founders.json (hunk 13 above)
Comment (tibetsprague) [reply migrations/founders.json]: yeah these should all be public keys, on ganache too. the private keys won't work

Comment (dkent600) [reply migrations/founders.json]: changed to addresses

File: migrations/2_deploy_organization.js (hunk 8 above)
Comment (dkent600) [reply migrations/2_deploy_organization.js]: Going with Genesis for the demo tomorrow

**Pull request merged**